### DISCUSSION TITLE
Deprecation of macOS Intel-based machines

### DISCUSSION THREAD

**Discussion by helinanever** [codemagic-ci-cd/codemagic-docs#2619] [Announcements]
Starting **September 1, 2024**, we are removing Intel-based Mac Pro machines and transitioning to Apple silicon-based hardware for all macOS builds. 

In the coming months, we will conduct brownouts to phase out Intel-based machines, see the brownouts schedule below. Builds still using Mac Pro machines after the EOL date will fail with an appropriate error message.

## Why are we making this change?

Apple has stopped using Intel processors in Mac computers, eventually making the Intel-based machines obsolete and requiring everyone to migrate to the Apple silicon machines. We will replace all Intel-based machines with Apple silicon hardware to provide better performance, up-to-date software, and the best possible support for all our users.

## What action is required?

This change affects you if you’re currently using any of the following:
- Xcode 11.x, 12.x, 13.x or 14.x
- Mac Pro (`mac_pro`) build machines

### Change instance type

Please update your build configuration to use one of the following available Apple silicon machines:
- `mac_mini_m1` – available for the Free tier and Pay-as-you-go users at the same per-minute price as Mac Pro
- `mac_mini_m2` – available for users on annual and custom plans

Please note that Android emulators are unavailable on Apple silicon machines due to the Apple Virtualization Framework not supporting nested virtualization. If you need the Android emulator, please switch to a Linux machine (`linux_x2`) instead. 

### Update Xcode version

The available Xcode versions on Apple silicon machines start from Xcode 15.0, the minimum requirement for App Store submissions. Please update your projects to use Xcode 15 or later. You may need to make additional changes if you use other pre-installed software. See the list of software preinstalled on our Apple silicon machines [here](https://docs.codemagic.io/specs/versions-macos/).

If you have any questions or require assistance migrating your workflows, please contact us via the in-app chat widget.

## Brownouts schedule and EOL

During the brownout periods, all builds using Intel-based macOS machines will fail. Please update your workflows as soon as possible to prevent any disruption to your builds.
- July 11, 2024, 06:00 -> 14:00 UTC 
- August 1, 2024, 04:00 -> 18:00 UTC
- August 22, 2024, 02:00 -> 22:00 UTC
- August 26, 2024, 00:00 -> 00:00 UTC (24 hours)
- September 1, 2024 – EOL. Deprecated machines are removed and become unavailable.

**Comment by bahag-chandrana** [comment]
This should have been announced in the website as well. We are now blocked with our pipelines 😞

**Reply by icarusdust** [reply]
Hi @bahag-chandrana, sorry to hear that it affected your workflow pipelines. Please note that we announced it on the [status page](https://status.codemagic.io/) as well. You can subscribe to the status page updates by clicking the `Get updates` button at the top right corner. 
To unblock your pipelines, try changing the `instance_type` to `mac_mini_m1` in the build configurations, if you are facing some other issues, then please let me know, I am sure that we can resolve them, so your pipelines can get back to normal.

**Reply by icarusdust** [reply]
We are communicating by sending out separate emails as well.

**Reply by bahag-chandrana** [reply]
Hi @icarusdust, Thanks for the response. Honestly i don't look at status pages unless there are some issues. Everyday we check our builds, there i didn't see any announcement. I remember we had announcement regarding xcode deprecations in our codemagic app view. 

My only feedback announcements should be where people actually work with the tool. Even emails are good as I check it on every business day. The more the merrier for breaking announcements 😄 

I have already tried linux machines and we had issues with our test runs, which I suspect is due to the simulator difference. We can try to fix that and get in touch, if we are unable to port to linux.

**Reply by icarusdust** [reply]
Thanks for the feedback @bahag-chandrana, it is appreciated. As for email communication, looks like you did not receive it, I will make sure that your email address is included in the list for future updates.

Sounds good, let me know how it goes with Linux instances, you can also use the chat widget in the UI in case of any questions.

**Comment by matthew-shin-hs** [comment]
It sounds like with macOS 15, there will be [nested virtualization](https://discussions.apple.com/thread/255314811?sortBy=best). Are there plans to support Android emulators with the Apple Silicon devices in the near future after deprecating Intel-based machines?

**Reply by mikhail-tokarev** [reply]
hey @matthew-shin-hs 
Yes, we are also looking forward and waiting for Apple to make nested virtualisation available again for developers. As soon as there are no restrictions from macOS or the hardware side, we will update our stack to support it. Thank you for your interest!

**Comment by Reedyuk** [comment]
Some of our workflows are using the intel based machines due to virtualisation of emulators, I assume the issue is still present that the Apple Silicon Codemagic machines don't support virtualisation, if this is the case, what is the proposed solution when you stop allowing the use of intel based machines?

**Reply by dtrdic** [reply]
Hello @Reedyuk It's important to note that this isn't a limitation specific to Codemagic, but rather a limitation from Apple. Apple Silicon machines don't support Android emulators due to their Virtualization Framework lacking nested virtualization capabilities.

In this case, we recommend using a Linux instance to run Android emulators. You can find detailed emulator specifications [here](https://docs.codemagic.io/specs/versions-linux/#android-emulators)

Although there are [some indications](https://discussions.apple.com/thread/255314811?sortBy=best) that nested virtualisation will be implemented in the future, and once there are no restrictions from macOS or the hardware side, we will update our stack to support it

**Reply by Reedyuk** [reply]
Just a small suggestion, would codemagic be open to the idea of supporting multiple virtual machines in a single workflow?(github call it a matrix) for example, We could then use Apple silicon all the way up to our integration tests/ui tests, then switch over to a linux machine for the android items.
The situation to flip flop between machines is because we are using KMP, so we want to compile for ios + android, perform unit tests, then perform integration and then ui tests.

**Reply by dtrdic** [reply]
Thank you for the feedback @Reedyuk i'll mark this suggestion as a feature request, so the team can take a look at it for future improvements.
But in any case if you want to use Android emulators in macOS M1 workflows, you can also use third-party service integrations such as emulator.wtf, Katalon, Firebase, or AWS Device Farm. Please check [this documentation](https://docs.codemagic.io/yaml-testing/emulator-wtf/) on how to run tests using emulators from emulator.wtf

**Reply by Reedyuk** [reply]
We may have no choice to use another third party for the device testing, but one of the reasons we chose to use and integrate codemagic was because of its emulator support, for this to be taken away(fairly short notice too), will mean additional cost dev time/alternative service, which we have not forecasted.

**Reply by dtrdic** [reply]
You can see some preliminary information about an API in development regarding the nested virtualization feature [here](https://developer.apple.com/documentation/virtualization/vzgenericplatformconfiguration/4360552-isnestedvirtualizationenabled?changes=latest_minor). I'm sure this will be released really soon and we are monitoring it closely

**Reply by Reedyuk** [reply]
excuse my ignorance but could you not wait until this is supported then? before switching off support for intel macs?

**Reply by dtrdic** [reply]
I understand your frustration. We announced the deprecation timeline 2 weeks ago, and Apple's recent announcement about supporting Nested Virtualization came just 5 days ago. Rest assured, we will update our system as soon as Apple rolls out these enhancements
